Fixed lag spike caused by unloading tileentities

CreativeMD opened the pull request:
Hey there,

I'm the author of LittleTiles/CreativeCore and recently I found out that Minecraft has a huge lag spike during unloading chunks if the world contains a lot of tileentities.

**How did I find it?**
Although the performance of LittleTiles was not bad there used to be quite horrible lag spikes, which occurred rather frequently while the player was moving. It became more and more unbearable. So I made several attempts to fix it, but unfortunately without any success. I couldn't find anything in my code.
A few months later I tried it again, but this time I was ignoring my code, since I suspected it to be an issue caused by Minecraft itself.

**The issue itself**
Eventually I found that the lag spike was not caused by loading new chunks, but instead while unloading the old ones. After some further investigations I saw those two suspicious looking lines in `World.updateEntities()`:
```java
this.tickableTileEntities.removeAll(this.tileEntitiesToBeRemoved);
this.loadedTileEntityList.removeAll(this.tileEntitiesToBeRemoved);
```
So I debugged it and ... yes ... wow. There were 20,000 tileentities in `tickableTileEntities` and `loadedTileEntityList` while `tileEntitiesToBeRemoved` had a size of 3,000. Stepping over it using the debugger took quite a long time.

I finally had found the issue. It's not caused by the amount of traffic or the complexity of tileentities, but caused by the amount of tileentities no matter if there are furnaces or more complicated ones.

This video demonstrates this issue: https://www.youtube.com/watch?v=vTCJbp45pWM

I also reported this issue to Mojang: https://bugs.mojang.com/browse/MC-119686

**How did i fix it?**
I replaced those two lists with `HashMap<ChunkPos, List<TileEntity>>`. So once a chunk will unload one key will be removed instead of iterating through the whole list several times. This completely solved the lag spikes (as shown in the video).

I implemented it in CreativeCore v1.8 and eventually decided to create this pr.

**About this PR**
I changed `tickableTileEntities` and `loadedTileEntityList` from an ArrayList to a `ChunkedTileEntityList`, which basically saves all tileentities in a `HashMap<ChunkPos, List<TileEntity>>`. Furthermore I added another list to the world (`tileEntitiesChunkToBeRemoved`), which contains all chunk positions of unloaded chunks. During each tick those will be removed from `tickableTileEntities` and `loadedTileEntityList`.
It's still possible unloading a tileentity using the old way, but I changed it for the chunk. A chunk will now call `markChunkForRemoval` instead of calling `markTileEntityForRemoval` (func_147457_a) for each tileentity.

If there is anything to complain about my naming or the formatting style, just let me know.

**Possible mod conflicts**
Since CreativeCore changed it already I can ensure that there is currently only one mod it is conflicting with https://github.com/RootsTeam/Embers/issues/237.
As long the iterator is used there are no changes required.

**Last words**
This issue must have existed for several years now. It probably caused a lot of headaches and definitely ruined the experience for many players, especially if you have build a lot of machines or other stuff close together. So I really hope this improves our modding situation. It affects the server in the same way as the client, only that it might be worse for the server (more loaded tileentities).

I'm looking forward to smoother gameplay and hope this PR will be merged soon. Once that is done I will create another one for 1.11.2.

**UPDATE**
All the indexed methods have been implemented. There are bad at performance, but the game will no longer crash of a mod is using them (like Embers).

In Regards
CreativeMD

CreativeMD committed:
Fixed lag spike caused by unloading tileentities

kashike commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -1,6 +1,18 @@
+ import net.minecraft.util.SoundEvent;
+ import net.minecraft.util.math.AxisAlignedBB;
+ import net.minecraft.util.math.BlockPos;
++import net.minecraft.util.math.ChunkPos;
```
do not import

kashike commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -1,6 +1,18 @@
+ import net.minecraft.world.storage.WorldInfo;
+ import net.minecraft.world.storage.WorldSavedData;
+ import net.minecraft.world.storage.loot.LootTableManager;
++import net.minecraftforge.common.util.ChunkedTileEntityList;
```
do not import

kashike commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,213 @@
+package net.minecraftforge.common.util;
```
Missing license header.

liach replied:
Fixed already

kashike commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,213 @@
+import net.minecraft.world.World;
+import net.minecraft.world.chunk.Chunk;
+
+public class ChunkedTileEntityList {
```
formatting

kashike commented:
See #3872.

ryantheleach commented:
Does the iteration order matter?

CreativeMD commented:
The iteration inside a chunk will be sorted as usual, only the order of the chunks will not be the same, i don't know if that is important.

CreativeMD committed:
Fixed formatting issues

williewillus commented:
You might want to have a read through https://github.com/MinecraftForge/MinecraftForge/pull/3872

liach commented:
I suggest using LinkedHashMap for iteration performance

mezz commented:
This is pretty much what mojang should do, but unfortunately we can't make this type of breaking change in Forge.
As other have said, please see the nearly identical investigation and final conclusion here https://github.com/MinecraftForge/MinecraftForge/pull/3872#issuecomment-303001598

CreativeMD committed:
Implemented Collection

CreativeMD commented:
I forget to implement the collection interface, so now it shouldn't crash if the iterator is used.

@liach I'm not really sure about that one. Insertion will be slower, iteration might be a bit faster.
@mezz As I said already, CreativeCore does those changes using asm already, without any sort of big problems. Embers is the only mod which iterates through the list using the `get(int)` method and as you can read in the issue itself he will eventually try to change it.

I'm not the one who should decide this, but I don't see any major issues (hashcodes are not needed). Basically it's just a bunch of lists glued together, so the iteration would be the only potential problem.

Cybermaxke commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,283 @@
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+
+public class ChunkedTileEntityList implements Collection<TileEntity>
```
Shouldn't this class implement `List`?

CreativeMD replied:
No, because implementing methods like `get(int)`/`remove(int)` would be horrible performance wise. People should be forced to use the iterator.

Cybermaxke replied:
Maybe it is better to just deprecate those methods and throw a unsupported operation exception? Right now you just change the field type, while devs may still expect a `List`.

CreativeMD replied:
Already thought about that one, but it wouldn't make a difference in my most cases. It would only change the exception which will be thrown once somebody uses those methods (like get(int), remove(int)).

The only case where it would make a difference would be if a mod casts it to a list (maybe to store it somewhere else). That is rather unlikely, but not impossible. So I'm not really sure about it.

I will leave it for now as it is, to keep it better organized. Maybe someone else knows if this is necessary or not.

tomelfring replied:
I think a `List` is more preferred as it maintains ~~(binary)~~ compatibility. Anything that is declared as public might be used by some mod, so best to play safe and let it be a List.

marvin-roesch replied:
Making it a `List` will maintain *source* compatibility, I think. Breaking binary compatibility is not too bad, since mods can just recompile against the new version, but breaking compatibility with existing code like this currently does is not an option, I'd say.

CreativeMD replied:
Ok, looks like java is a little bit weird and refers to the list interface instead of the parent one. I will change it to a list real quick.

liach commented:
hashmap is super slow at iterations! o(n) vs. o(n+buckets)

williewillus commented:
Uhhhh but the bucket count is bounded by n if your load factor is anywhere less than 1, so it just ends up O(2n) = O(n). I would hardly be worried about that.

@CreativeMD as it stands lex rejected the last attempt at this so you should really state what makes this faster or safer or better than the last one, so you don't have to waste time reimplementing.

CreativeMD committed:
Replace HashMap with LinkedHashMap

CreativeMD commented:
@williewillus The PR is quite different from mine. @mezz replaced the list with a LinkedHashSet, which obviously caused a lot of problems (hashcode() causing issues). As a result he came up with another idea:
> These big lists should be broken up by chunk.

Which is basically what I did (without knowing anything about the PR before).

I said everything from my end. It is working and only causes issues with mods which don't use the iterator. Now it's up to Lex to decide whether a new version of forge should break a few mods or if this lag spike will continue to exist.

tomelfring commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,284 @@
+		return content.entrySet();
+	}
+	
+	public boolean containsAll(Collection<?> tileEntities)
```
Missing @ Override

CreativeMD committed:
Added override annotation
Implemented List instead of Collection

liach commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -17,7 +17,18 @@
++    public final net.minecraftforge.common.util.ChunkedTileEntityList field_175730_i = new net.minecraftforge.common.util.ChunkedTileEntityList();
+     private final List<TileEntity> field_147484_a = Lists.<TileEntity>newArrayList();
+     private final List<TileEntity> field_147483_b = Lists.<TileEntity>newArrayList();
++    public List<net.minecraft.util.math.ChunkPos> tileEntitiesChunkToBeRemoved = Lists.<net.minecraft.util.math.ChunkPos>newArrayList();
```
I suggest making this
```java
public final Collection<net.minecraft.util.math.ChunkPos> tileEntitiesChunkToBeRemoved = new ArrayDeque<>();
```
It's java 8 so we can use diamonds. Using an array double queue makes both addition and iteration fast.

N247S replied:
adding and iterations arent the main problem, its mostly the removing part. Although some alternatives you suggested may or may not improve on the first two operations, they may influence removing in a more negative way (which is the main problem atm).

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+{
+	
+	private static ChunkPos getChunkPos(TileEntity te)
+	{
```
ewww tabs!

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+		return false;
+	}
+	
+	public void add(TileEntity[] tileEntities)
```
Make it `add(TileEntity... tileEntities)` for easier use.

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	@Override
+	public boolean addAll(Collection<? extends TileEntity> tileEntities)
+	{
+		for (Iterator<? extends TileEntity> iterator = tileEntities.iterator(); iterator.hasNext();) {
```
Why not a for each loop?

CreativeMD replied:
Basically the same, i just got used to it, but i will change it for you.

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	public ChunkedTileEntityList(ChunkedTileEntityList list)
+	{
+		this();
+		for (Iterator<Entry<ChunkPos, List<TileEntity>>> iterator = list.entrySet().iterator(); iterator.hasNext();) {
```
Use a for each loop
`for (Entry<ChunkPos, List<TileEntity>> entry : list.entrySet())`
and the left braces!

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	{
+		ChunkPos pos = getChunkPos(te);
+		List<TileEntity> list = getChunkTEList(pos);
+		if(list == null)
```
use `computeIfAbsent`

CreativeMD replied:
Would there be any benefits?

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	@Override
+	public Iterator<TileEntity> iterator()
+	{
+		return new Iterator<TileEntity>() {
```
How about using guava utilities here?

CreativeMD replied:
Never used them, i doubt it it important.

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	
+	@Override
+	public boolean isEmpty() {
+		return totalCount() > 0;
```
Hmm obvious error here

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	}
+
+	@Override
+	public <T> T[] toArray(T[] paramArrayOfT) {
```
Replace `<T>` with `TileEntity`

tomelfring replied:
It's an overridden method of collection.

marvin-roesch replied:
That isn't valid, I think. Due to the way Java's arrays etc work, a list needs that generic method in order for the runtime types to work out properly. Refining the return type doesn't work here for that reason.

liach replied:
No but the class implements `List<TileEntity>` you silly guys!

liach replied:
What other types of array can you really cast into?

marvin-roesch replied:
That doesn't change anything about the *method* being generic in `T`. The `T` here isn't the same as the one for the `List`.

N247S replied:
'T'  is a generic type of the method. you may replace it with 'Object' (cause of type ereasure), but replacing it with anything else will break the code. (try it out, you'll see for yourself!)

marvin-roesch replied:
It's simply a requirement by the `List` interface. Unlike other methods which return a specific (i.e. non-generic) type (take any of the Item methods that return `this`), you can't refine the return type (e.g. from `Item` to `MyItem`) in this instance.

liach replied:
Ok. Let's go to see other follies like that `isEmpty` one!

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+	@Override
+	public <T> T[] toArray(T[] paramArrayOfT) {
+		if(paramArrayOfT.length < totalCount())
+			paramArrayOfT = (T[]) java.lang.reflect.Array.newInstance(paramArrayOfT.getClass().getComponentType(), totalCount());
```
Use imports instead of qualified class names

liach commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -458,14 +469,26 @@
++        
++        if (!this.tileEntitiesChunkToBeRemoved.isEmpty())
++        {
++    		for (Iterator<net.minecraft.util.math.ChunkPos> iterator2 = tileEntitiesChunkToBeRemoved.iterator(); iterator2.hasNext();) {
```
1. braces
2. `for (ChunkPos pos : tileEntitiesChunkToBeRemoved)`
I believe World class already imported ChunkPos, is that true?

CreativeMD replied:
No it's not imported, somebody else said i should remove all of my imports.

liach commented:
I agree with this fix, but the code is problematic.

N247S commented:
the code can be improved a bit. But its base is actually ridiculously simple and effective. I seriously wonder why this hasnt been done before by either the MC-developers or by Forge itself (especially Forge)

Chocohead commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,343 @@
+		Object[] array = new Object[totalCount()];
+		int i = 0;
+		for (List<TileEntity> values : content.values()) {
+			for (int j = 0; j < values.size(); j++) {
```
Why not something like
```java
for (TileEntity te : values)
    array[i++] = te;
```

N247S replied:
that would work, but why create the confusion? I mean, is there any benifit of performance involved? To my knoweledge it will create an Iterator which will slow things down beside creating confusion.

CreativeMD committed:
A lot of small fixes/enhancements

CreativeMD commented:
@liach the `isEmpty()` issue was the only mistake so far. The rest of the code was working just fine. It's just a matter of preferences.

liach commented on src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Override
+    public Iterator<TileEntity> iterator()
```
I guess you keep an iterable like this: https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Iterables.java#L499-L500
and you return the iterator of that iterable every time the iterator is called

CreativeMD replied:
Would there be any benefits from it?

mezz commented:
I don't think there is anything wrong with your fix, I think it looks fine.
The issue is that Forge does not accept *any* binary incompatible changes except in extreme circumstances like the game doesn't work at all or it's deleting worlds.

~~For something like this, we can't accept it until the breaking change window during the move to 1.13.
If we're really going to go ahead and break it, we shouldn't bother with implementing `List` at all since it's really limiting.~~
Talked to Lex, and we can't make a fundamental change like this, even in 1.13. Forge has to work with other systems and can't diverge from vanilla this way. We really need this fix from Mojang.

N247S commented:
@mezz So just out of curiousity, if the base implementation was an instance of 'List' (which doenst break binary compatibility), this would be fine?

CreativeMD committed:
Changed it to back to be a list

CreativeMD commented:
@mezz oh yes, sorry. Totally forgot that GETFIELD also requires a desc, I changed it back to list.

asiekierka commented:
There's an easier way to "fix" it which doesn't break binary compatibility: you can simply ensure that WorldServer.updateEntities calls a method to remove entities and tile entities even if no players are present on the server. Now, this does not make removals in any way faster, it just ensures they don't "pile up".

CreativeMD commented:
@asiekierka this fix does not break any binary compatibility, but will throw an exception as soon as `get(int)`, `remove(int)` or other similar methods are called (basically it is a collection now, not a list anymore).

asiekierka commented:
@CreativeMD It doesn't break binary compatibility but it has the same effect in practice: code (and you can't know if such code does not exist) which worked on the last recommended version of Forge for 1.12 will not work on the next recommended version of Forge for 1.12.

CreativeMD commented:
@asiekierka Only a few mods care about those lists anyway and even if they do it will only break those which don't use the iterator. It only affects the embers mod (other cases are not known so far).

williewillus commented:
"Only a few" is always more than you think

CreativeMD commented:
As i said earlier. CreativeCore does it already and embers is the only mod so far. This is not just speculation. It's been tested for about two weeks now.

liach commented:
well, I've never heard of creativecore and embers mods before

covers1624 commented:
Yeah, the argument of "My coremod is doing this already should be fine" isn't really valid.

CreativeMD commented:
Guys please ... there is no point in all this. If it's not worth it in your eyes, close this PR. I can only provide the information I have. I can't guarantee for anything.

liach commented:
@CreativeMD You should close it yourself. You need to write better code.

N247S commented:
@liach What a comment..., so you can do it perfect right?

asiekierka commented:
@N247S I can at least propose a way to fix the complaint I've had:

It's easy to add support for get/set/remove: as the iteration order is deterministic in a LinkedHashMap, for get(i) you just subtract each List's size until you get to a List which can contain said i; you can do the same for set, remove, etc. indexOf can be done in the same manner: you add the sizes until you find a List which contains the item in question.

CreativeMD commented:
@asiekierka I'm not stupid. Of course it is possible to implement that, but as i said already, it's horrible performance wise. If you will implement it people will use it.

Sorry if i sound rude, but we are spinning in circles at the moment.

marvin-roesch commented:
First of all, may I point out that this seems like the perfect usecase for Guava's `Multimap` to me, specifically a [`LinkedListMultimap`](https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/LinkedListMultimap.html)?
Secondly, the `List` methods should be implemented, no matter how bad they perform. Keeping source compatibility is more important (imo) than optimising these operations. If people notice performance drops with specific mods, issues can be reported to them and they can be made aware of these implementation details.

asiekierka commented:
@CreativeMD The thing is, you could implement it now and then break the methods in 1.13, when compatibility with existing mods is no longer a concern.

CreativeMD commented:
Yes okay, that's a good point. I will implement it. Sorry for my first reaction (might be caused by this atmosphere here).

@PaleoCrafter I would rather pick ArrayListMultimap instead (performance wise), but that one doesn't make anything difference. It's basically the same thing as i'm doing right now. So i'm not sure if it is necessary to switch to it.

marvin-roesch commented:
The interface for Multimaps is nicer, imo, you get some guarantees you have to take care of manually otherwise. I suggested LinkedListMultimap because it's semantically equivalent to your current implementation (iteration order = insertion order for keys *and* values).

CreativeMD commented:
ArrayListMultimap would sort it per chunk, while LinkedListMultimap would keep the order of insertion. I don't know which sorting is is better, but it looks like LinkedListMultimap can become really handy. The only question i have, how good is it performance wise (especially RAM usage)?

EDIT
It might even us hashcodes of the values (which is a problem). So I rather leave it out for now.

mezz commented:
@liach that is unacceptably rude. Consider this a warning.

liach commented:
You guys need to pick one between two: compability or performance. An attempt to keep both one of them will result in the loss of both.

liach commented:
mezz, I say that because I see some critical errors like the isEmpty one (how could no one else noticed that and all the formatting issues!). You should go to the SpongePowered repos to see what the strict rules and good code are. However, this is simply a forum thread because this pull is not really getting accepted.

williewillus commented:
As I have mentioned before, no matter trivial or "stupid" an issue seems to you, you need to state it respectfully, nothing more. Adding rude fluff only serves to deter people who actually want to contribute but are afraid of being flamed.

mezz commented:
This may be accepted if it is done carefully.

- There must be no breaking changes (i.e. it has implement `List` completely). Some methods will be slower than ideal, but do what you can.

- You should show performance before and after, using the test setup from the vanilla bug here:  
 https://bugs.mojang.com/browse/MC-117075

Once that's complete I can go over style issues with you. Generally there should be no tabs, any new fields should be below the existing ones, minimize patch sizes, picky stuff like that.

liach commented:
Looking forward to seeing the pull request getting accepted.

CreativeMD committed:
Added support for indexed methods (N247S)
Improved performance for totalCount
Added fail-fast concurrent protection to the iterator

CreativeMD commented:
All indexed methods are added now (only `listIterator()` and `subList()` are missing). They are marked as deprecated, hoping modders will stop using them once they will see it.

N247S commented:
We could implement the `SubList()` and `ListIterator()` this though would it be necessary? It will break the `List` interface contract as absolute index support cannot be granted (especially with the indexed `add()`/`set()` methods from both `SubList` and `ListIterator`).

Either we ignore indices and call its non-indexed counterpart (for the ListIterator not possible), or we could check if the current (absolute) position is valid (since a `Tilenentity` should be in the right 'chunkList'), and if its not the case throw an Exception.

SquidDev commented:
Could you not extend `AbstractList`? - this way you get `addAll`, `removeAll`, `retainAll`, `subList` and `listIterator` for free. The last two may not have the most efficient implementation given the structure of `ChunkedTileEntityList` but would be better than throwing an exception.

N247S commented:
@SquidDev It would simply break this implementation, as it does not correctly takes care modcount, empty chunklists etc, besides overal performance is quit bad. Appart from that, all the methods are already implemented (even the ListIterator and SubList is ready for commiting, only the indices part might break more than it fixes, which is why we wait for the communities opinion about this matter).

Mathe172 commented:
Another possibility (that is 100% backwards compatible):
Implement your own (fully functional) LinkedList (`implements List<TileEntity>`) on top of `AbstractSequentialList`. Additionally, store a `Map<ChunkPos, Map<TileEntity, Node>>` in it.

Now you can implement `remove(Chunk)` and `remove(TileEntity)` pretty efficiently:
* `remove(Chunk)`: Just cut out the nodes from your LinkedList. While not as cheap as simply forgetting part of the list, it should still be much more efficient than the current implementation.
* `remove(TileEntity)`: Get the `Node` from the Map and remove it.

HenryLoenwind commented:
![image](https://user-images.githubusercontent.com/1485873/28745128-c7388eca-7471-11e7-9b6e-44c64b58f786.png)
like this, Mathe172?

CreativeMD commented:
@Mathe172 There are a lot of ways to improve performance for those indexed methods. The LinkedListMultimap (suggested by @PaleoCrafter) is just one example.

The downside of this is extra RAM usage and decreased insertion performance. It's again a matter of choice. Right now the current implementation is fully functional (leaving out the `listIterator()` and `subList()`), but the index iteration is not as fast as it used to be.
So the question is (to the Forge developers), should we improve performance for something we want to remove later on (in 1.13) or should we keep as it is now.

Please  also keep in mind, that we are not allowed to use the `hashcode()` of a tileentity, which limits our possibilities of flat data-structures.

PhiProven commented:
Your current implementation of `add` (indexed version) and `set` aren't conform with the `List` specification.

`LinkedListMultimap` doesn't implement the `List` interface, it had to be added manually. This would have the same problem with `add` and `set` (and maybe some other indexed methods).

This problem is not present in @Mathe172's solution, as elements can be added at arbitrary positions in the linked list; the index is not restricted by the chunk position.

I don't know if it is really necessary to have an implementation of `add` and `set` that is conform with the `List` specification, though I would rather throw an `UnsupportedOperationException` than redefining the meaning of the index...

Mathe172 commented:
@HenryLoenwind Like this:
![telist2](https://user-images.githubusercontent.com/3540742/28745671-2d1aa4c0-747d-11e7-8656-78ee64c4727f.png)
[Edit: Fixed diagram]

Mathe172 commented:
We implemented a fully working version of the List [here](https://github.com/Mathe172/MinecraftForge/blob/93049a9af4bcaa4f3820367a7745ee2a1b33fbec/src/main/java/net/minecraftforge/common/util/TEList.java). The previous suggestion had some issues with the `remove(Object)` method (and the fact that we can't call `TileEntity.hashCode()`), so we switched to a quadruply linked list, with a second layer of lists that join the nodes belonging to a specific chunk:
![telist3](https://user-images.githubusercontent.com/3540742/28757322-8490d586-7580-11e7-8d1a-e53ca903f134.png)

williewillus commented:
I feel like the solutions are getting so complex we're probably better off bugging Mojang to fix it themselves by just turning it into a set.

marvin-roesch commented:
@Mathe172, if I understand it correctly, that particular implementation would only be required if we wanted efficient implementations of the indexed `List` methods, right? If that's the case, I'm still of the opinion that we shouldn't care much about their performance and avoid such complex data structures. Teaching a few modders how to use this implementation "properly" is probably easier.

Mathe172 commented:
@PaleoCrafter: Indexed operations are not optimized at all in our implementation. But something like that is required if you want to properly implement all methods of the list interface. As explained above, the implementation that groups TileEntities by Chunk is unable to properly implement `add` for example.

CreativeMD commented:
@Mathe172 Yes, basically there is no sorting, so it will just be ignored, which isn't a problem in my eyes. However this can potentially cause weird behavior when working with the listIterator().

Mathe172 commented:
@CreativeMD Well, you break the contract of the `List` interface - in my opinion this is even worse than throwing an `UnsupportedOperationException`, as the "weird behavior" will most likely lead to untraceable bugs somewhere.

CreativeMD commented:
@Mathe172 That's what we are currently doing. `listIterator()` and `subList()` are currently throwing an exception. @N247S asked if we should implement those methods although they might have a weird behavior.

PhiProven commented:
The current implementation of `add(int, TileEntity)` and `set(int, TileEntity)` also break the contract of `List`.
I think it doesn't make sense to implement them, but leave out `listIterator` and `subList`. I would either throw an exception for all of them or implement them properly (which requires something like the solution @Mathe172 and I provided, I think).

CreativeMD commented:
It does what it is supposed to do (will add the given tileentity/ replace it with the old one), but it will ignore sorting, which should be fine in almost every case.

Actually, I'm not really sure if sorting is required by any existing mod. I think those lists are most commonly used to iterator through and maybe even to remove/add or search for an tileentity. All of those things are working just fine. So if it is not necessary we should leave any additionaly data-structure, to increase performance. Again it's a decision which has to be taken by a forge developer.

williewillus commented:
The "decision which has to be taken by a forge developer" was that this needs to be fixed by vanilla, see #3872.

liach commented:
No. We are going to fix it here, as mezz implied that it has a chance to get accepted.

N247S commented:
As an extra we thought it would be a good thing to provide a sumary of where things stand at this moment. As we see that there are a lot of good arguments given for the different choices (some suggestions are even reapeated several times). @CreativeMD and I have had several long disccusions about what the best course of action is for implementing this system. So we would like to go over all the possibilities.


**The Problem**
First of all its a good thing to confirm that the TileEntity '`List`' is actually used as a `Collection` by MC itself (e.g. no indexed acces, only enhanced for-loop iteration and `remove(Object)`/`add(TileEntity)` calls). The most basic implementation they have chosen is the `ArrayList` (saved as plainly a `List`). In the early days of MC this wasn't a problem as chests and furnaces were basically the only TileEntity implemented at that time. With the current MC versions there are already a lot more (e.g. Comparator, Daylight sensor, hopper, chulkerbox etc.), and with the growing modding community its becomming more and more important that the best performance is granted while working with TileEntities.

The main problem is that for non-indexed acces (to an non-hashed/non-sorted list implementation), the full collection (in the worst case) have to be iterated in order to obtain the right element to perform the operation on. There has been an earlier attempt by using a hash-based collection, though it gave a lot of problems (as seen in this post #3872).


**The solution(s)**
We tried to tackle the problem by storing the TileEntities by Chunk (to literly chunk the data), which would make non-indexed acces less resourcefull (keep in mind, non-indexed acces should be the main focus here, as described above!). The only catch is that indexed-acces cannot be fully maintained while trying to provide optimal performance.

The possible options for the indexed-operations are as followow:

- No  implementation at all (e.g. throw an exception) as indexed acces shouldnt be necessary, and is nontheless a bad performance choice.
- Implement the full indexed acces compatibility, by maintining adding/insertion order through for example a linked list (giving indexed-acces a liniar timecomplexity). non-indexed acces can still be optimized through the current implementation, though it will affect the overal performance in a negative way.
- change the concept of indexing by breaking the `List` contract. In here we have mainly 2 choices. Both of these options are able to use the chunking to enhance indexed-acces performance compared to a array/linked List.
    1. Treat the index as a z-index. With this we mean (like we implemented currently) that by an index of 0, the TileEntity will be placed at the start of the List of its own chunk, and will therfore be the first upcomming in the iterator (when arriving to that chunk). An index of `TileEntityList.size()` means it will be placed at the end of the List of its own chunk.
    2. Allow absolute indexing, but throw an exception when a user tries to place an TileEntity at the wrong chunk list. Additionally we could provide methods to get the `chunkOffset` and `Chunksize` to help out with indexed acces.

**Disclaimer!**
The reason this discussion exist is soly based on the fact that multiple mods uses the index-methods. @CreativeMD and I have taken the time to search through a lot of Github repositories in which TileEntityList acces is vissible. We found a lot of indexed based loops to iterate all the TileEntities. In other words, basically all index based acces is in context of a non-enhanced for loop, making the importance support for other indexed acces (`add`/`set`/`remove`/`subList`/`ListIterator`) verry small! (nontheless we currently have added an implementation for `add`/`set` and `remove`)

In our oppinion we should protect the modder against unefficient TileEntity acces by disabling index-acces completly rather than implementing a semi-working system, or by decreasing the performance which basically means we could beter leave things as they are atm.


**Conclusion**
Basically everyone has a valid point by saying there are alternatives, though at this moment we have to wait on the ForgeDevelopers to make a decision about what course of action will be the best. maybe keep the contract for v1.12, but break it for v1.13. Lets wait and see!

ryantheleach commented:
>Implement the full indexed acces compatibility, by maintining adding/insertion order through for example a linked list (giving indexed-acces a liniar timecomplexity). non-indexed acces can still be optimized through the current implementation, though it will affect the overal performance in a negative way.

That would be my pick, do the fix with the most compatibility, then check to see if *in practice* if the performance issues are worth the trade off.

Trying to guess without a design to measure against is crazy.

mezz commented:
I agree it doesn't make sense to optimize this or making compromises without a baseline.
We should test the performance of the "works perfectly" [`TEList`](https://github.com/Mathe172/MinecraftForge/blob/93049a9af4bcaa4f3820367a7745ee2a1b33fbec/src/main/java/net/minecraftforge/common/util/TEList.java) implementation.
It's not as performant as possible but it will almost certainly beat `ArrayList` by a very large margin.
If you find it's still very slow or takes 100x more memory or something terrible, then we can look at the compromised solutions.

Mathe172 commented:
We noticed that our TEList has one possible issue: If someone manages to override `TileEntity.equals()` in such a way that it can return `true` for TEs in different chunks, the current implementation of `remove(Object)` is wrong, as it only searches through the TEs belonging to the same chunk as the one to remove. 

There are three possible solutions:
* Assume `equal` TEs are in the same chunk. This is the current implementation 
* Assume `equal` TEs are at the same position. This might allow for a slightly more efficient implementation, though we'd have to look into that. 
* Assume the worst and fix `remove()`. The fix would simply make the complexity`remove()` linear in the total length as opposed to linear in the amount of TEs from the same chunk.

Which one would you prefer?

diesieben07 commented:
Why do you want to check `equals` at all? Use `==`.

Mathe172 commented:
You need to call `equals`, see [here](https://docs.oracle.com/javase/7/docs/api/java/util/List.html#remove(java.lang.Object))

Mathe172 commented:
(We could of course also implement it using `==`, but remember that the `ArrayList` used in vanilla uses `equals`.)

mezz commented:
Overriding `TileEntity#equals` is a bad idea but I've seen it before...
Since `TileEntity` has a `pos` field I think it's reasonably correct to assume that `equal` TEs are at the same position.
If you can't assume that, I don't think you can assume anything about their location with `equals`.

ryantheleach commented:
>"Assume the worst and fix remove(). The fix would simply make the complexity remove() linear in the total length as opposed to linear in the amount of TEs from the same chunk."

My first thoughts were to do this, but check the expected position/chunk _first_ then it's only going to be linear in the worst case scenario of the tile entity not existing in the list.

How often does remove() get called when the entity isn't already in the list?


But it also has the side effect of violating this contract. 😢 

>Removes the **first** occurrence of the specified element from this list, if it is present (optional operation). If this list does not contain the element, it is unchanged. **More formally, removes the element with the lowest index i** such that (o==null ? get(i)==null : o.equals(get(i))) (if such an element exists). Returns true if this list contained the specified element (or equivalently, if this list changed as a result of the call).

Mathe172 commented:
Regarding the optimization we mentioned: Assuming that `TileEntity.equals` respects `pos`, it would be possible to implement something along the line of my first suggestion, where instead of a second level of linked lists, the node references are stored in a `Map<BlockPos,List<Node>>`:
* Code wise, this might be slightly longer
* `.remove(Object)` would be a lot faster (constant time vs linear time)
* `ListIterator.add(Object)` would be slower in case that a TE already exists at the same position (which should not be the case), otherwise it would be faster

Do you think that it's worth it to implement this? Given that Vanilla does in fact use `remove(Object)`, this might be worth the effort. The only other drawback of this approach is that the implementation of `List` is no longer perfectly following the documentation (as we're assuming something about `equals`).

liach commented:
It is wrong to check equals by position because tile entities are initialized without anything, not even a world.

Mathe172 commented:
I'm not quite sure that I understand what you mean: Which part of the TE's life cycle are you concerned about? If we can't assume that the TE has a constant and correct `pos` when it's added to the list, this whole thread is pretty pointless.

That being said, has anyone considered to just replace the `ArrayList` with a `LinkedList` and to reimplement the chunk unloading to iterate once over the list and remove anything with a matching `pos`?
Some remarks considering this:
* Code wise, this is massively simpler
* Speed wise, it reduces the complexity of chunk unloading from `#World_TEs * #Chunk_TEs` to `#World_TEs` (while TEList has `#Chunk_TEs` complexity)
* The only assumption you make is that any TE in the chunks TE list has a `pos` that is in that chunk (at the time of unloading).

howtonotwin commented:
@liach Why does position being uninitialized make comparing positions for `equals` invalid?
 Logically, a `null` world means the TE is not part of any world and does not exist anywhere. Two TEs that are same in all respects except that one actually exists and the other is floating in the Void can be said to be unequal. The problem is what the semantics of `equals` are for TEs, specifically whether two identical TEs that happen to be in different places are "equal" or not.

This problem is, in fact, moot, because any override of `equals` that *doesn't* check for position would break the current implementation of the TE list, which is a contract-compliant `ArrayList`, and be fixed in short order when someone noticed that breaking a block *here* deleted a stack of diamonds *there*. <sup><sup>(or something)</sup></sup>

I vote we do what @Mathe172 just suggested.

mezz commented:
`LinkedList` with a better algorithm for removing the TEs sounds good to me. It's still a massive improvement in time complexity over the current situation.

PhiProven commented:
Probably, we don't even need a `LinkedList`, `ArrayList` should be fine:
* `remove(Object)` already needs to traverse the list to find the correct element. In average it has to go through half of the list. If it is iterated to the end in order to copy the elements, this shouldn't really be worse than a factor of 2 in performance.
* `remove(Chunk)` already has to iterate over the whole list. If we use `ArrayList.removeIf(...)` the list is only copied once (not for every removed element). Hence copying should only have a small performance impact.
* Copying a single element should be fast (only a single instruction) while calling the virtual `equals` and the predicate should be more "expensive". Hence I think the performance overhead should be much less than the factor 2.

The `ArrayList` has the advantage that:
* It has better cache locality (This shouldn't really matter in Java, but may make up for the little performance overhead).
* Has smaller memory footprint (Shouldn't be importat if the list doesn't have millions of entries).
* Indexed operations remain fast.

mezz commented:
There are no significant downsides to using `LinkedList` over `ArrayList` in terms of compatibility so I don't think we should shy away from using it.

I think we have discussed this very fully, what we need now is an implementation and a benchmark. Since there haven't been changes here yet, I think I'll get started on one implementation.

mezz commented:
Tested out some implementations using the furnace setup from [MC-117075](https://bugs.mojang.com/browse/MC-117075)

Implementation  | Time (ms)
------ | -----------
`ArrayList.removeAll(tiles)` (Vanilla) | 4154
`ArrayList.removeIf(inChunks)` | 235
`LinkedList.removeIf(inChunks)` | 449

Seems pretty clear to me which one to go with, I'll submit a PR.

liach commented:
mezz's new fix is really neat indeed.

mezz commented:
Submitted the PR here https://github.com/MinecraftForge/MinecraftForge/pull/4281
